polkadot/node/core/pvf/prepare-worker/src/lib.rs

```diff
@@ -229,31 +234,74 @@ pub fn worker_entrypoint(
 					},
 				};
 
-				// SAFETY: new process is spawned within a single threaded process. This invariant
-				// is enforced by tests.
-				let result = match unsafe { nix::unistd::fork() } {
-					Err(errno) => Err(error_from_errno("fork", errno)),
-					Ok(ForkResult::Child) => {
-						// Dropping the stream closes the underlying socket. We want to make sure
-						// that the sandboxed child can't get any kind of information from the
-						// outside world. The only IPC it should be able to do is sending its
-						// response over the pipe.
-						drop(stream);
-						// Drop the read end so we don't have too many FDs open.
-						drop(pipe_reader);
-
-						handle_child_process(
-							pvf,
-							pipe_writer,
-							preparation_timeout,
-							prepare_job_kind,
-							executor_params,
-						)
-					},
-					Ok(ForkResult::Parent { child }) => {
-						// the read end will wait until all write ends have been closed,
-						// this drop is necessary to avoid deadlock
-						drop(pipe_writer);
+				cfg_if::cfg_if! {
+					if #[cfg(any(target_os = "linux"))] {
+						let stack = 1024 * 1024;
+						let mut stack: Vec<u8> = vec![0u8; stack];
+						let flags = CloneFlags::CLONE_NEWCGROUP
+						| CloneFlags::CLONE_NEWIPC
+						| CloneFlags::CLONE_NEWNET
+						| CloneFlags::CLONE_NEWNS
```

Comment:
Great observation! I'm going to add a check for the `CAP_SYS_ADMIN` before doing the clone. As the `unshare_user_namespace_and_change_root` is being used for the `CLONE_NEWUSER` and `CLONE_NEWNS`, I can remove both of them from here, right? Is that ok for you @mrcnski?

Reply:
I believe `CLONE_NEW_USER` should already give us `CAP_SYS_ADMIN`:

              This flag has the same effect as the [clone(2)](https://man7.org/linux/man-pages/man2/clone.2.html)
              CLONE_NEWUSER flag.  Unshare the user namespace, so that
              the calling process is moved into a new user namespace
              which is not shared with any previously existing process.
              As with the child process created by [clone(2)](https://man7.org/linux/man-pages/man2/clone.2.html) with the
              CLONE_NEWUSER flag, the caller obtains a full set of
              capabilities in the new namespace.

This is how we are able to use `CLONE_NEWNS`, which also requires `CAP_SYS_ADMIN`, in `unshare_user_namespace_and_change_root`.

But, indeed it is possible that `clone` with these flags may fail on some systems/configurations. We may need to add a new capability to `SecurityStatus`. We can test ahead of time whether this `clone` call is expected to succeed, like we do with landlock, seccomp etc. If the check fails, and the user is not running in secure mode, we would fall back to fork here. I can implement this part if it sounds good.

Reply:
> But, indeed it is possible that clone with these flags may fail on some systems/configurations. We may need to add a new capability to SecurityStatus.

Yes, I think we should treat these extra namespace flags as separate security features (like we do for the user namespace).

But why do we bother with switching the fork+unshare to clone? I don't think there's a big benefit in doing so (and looks like it's quite complicated to implement in a bulletproof manner)

Reply:
@alindima We sandbox the worker soon after it starts, with some exceptions, because it doesn't need much I/O functionality. It seems good to keep this, even though this process doesn't directly interface with untrusted code anymore. When spawning the job process, we can restrict it even further, and doing it at `clone` seems like the appropriate time for it.

But, perhaps we could move some of the flags to the `unshare` call in the parent worker. Probably the worker doesn't need most of these capabilities. Then this `clone` call could be redundant. I can test it when I have a moment.

> (and looks like it's quite complicated to implement in a bulletproof manner)

Why do you say this?

Reply:
I'm mainly saying that the other `unshare` call received special treatment by being added as a security feature, with more robust handling in case it's not supported. I'm just saying that these extra namespace flags should be handled similarly, to preserve compatibility and avoid errors

Reply:
> But, perhaps we could move some of the flags to the unshare call in the parent worker.

Punted on this because it could cause sudden loss of the unshare-and-pivot-root capability on some validators. It is enough for the `clone` call to have these flags to be reasonably restricted.

> I'm just saying that these extra namespace flags should be handled similarly, to preserve compatibility and avoid errors

Will do this next. Without the CI that koute suggested we don't know what are the requirements for the union of all these flags. I'm therefore inclined to make this clone-with-sandbox-flags capability optional for Secure Validator Mode. It's also not needed to reasonably sandbox networking and filesystem access (our main threat model is protecting validator keys), so seems fine not to enforce it (we would still print a warning if it's missing).